server/etcdserver/api/v3rpc/maintenance.go

```diff
@@ -87,6 +88,8 @@ func NewMaintenanceServer(s *etcdserver.EtcdServer) pb.MaintenanceServer {
 
 func (ms *maintenanceServer) Defragment(ctx context.Context, sr *pb.DefragmentRequest) (*pb.DefragmentResponse, error) {
 	ms.lg.Info("starting defragment")
+	ms.hn.StopServe("defrag is active")
+	defer ms.hn.StartServe()
```

Comment:
There are chances that with a write-intensive workload, it will take some time for the local node to apply the backlogged in-flights entries. 

But this PR should make etcd availability much better when defrag is active.